File: tools/form-executor/src/normalization.mjs
export function canonicalizeAnswer(fieldType, value) {
  const nfc = (text) => String(text).replace(/\r\n/g, '\n').replace(/\r/g, '\n').normalize('NFC');

  const canonicalDecimal = (input) => {
    if (typeof input === 'boolean') throw new Error('boolean is not a canonical number');
    let raw;
    if (typeof input === 'number') {
      if (!Number.isFinite(input)) throw new Error('number answers must be finite');
      raw = String(input);
    } else if (typeof input === 'string') {
      raw = input.trim();
    } else if (typeof input === 'bigint') {
      raw = input.toString();
    } else {
      throw new Error('number answers must be numeric or decimal string');
    }
    if (!raw) throw new Error('number answer must not be empty');
    const match = raw.match(/^([+-]?)(\d+)(?:\.(\d*))?(?:[eE]([+-]?\d+))?$/);
    if (!match) throw new Error('number answer is not a valid finite decimal');
    const negative = match[1] === '-';
    const integer = match[2];
    const fraction = match[3] || '';
    const exponent = Number(match[4] || '0') - fraction.length;
    if (!Number.isSafeInteger(exponent)) throw new Error('number exponent is outside canonical range');
    let digits = `${integer}${fraction}`.replace(/^0+/, '');
    if (!digits) return '0';

    let rendered;
    if (exponent >= 0) {
      rendered = digits + '0'.repeat(exponent);
    } else {
      const point = digits.length + exponent;
      if (point <= 0) rendered = `0.${'0'.repeat(-point)}${digits}`;
      else rendered = `${digits.slice(0, point)}.${digits.slice(point)}`;
      rendered = rendered.replace(/0+$/, '').replace(/\.$/, '');
    }
    rendered = rendered.replace(/^0+(?=\d)/, '');
    return negative && rendered !== '0' ? `-${rendered}` : rendered;
  };

  const canonicalDate = (input) => {
    if (typeof input !== 'string') throw new Error('browser date answer must be ISO string');
    const raw = input.trim();
    if (!/^\d{4}-\d{2}-\d{2}$/.test(raw)) throw new Error('date answer must use ISO YYYY-MM-DD');
    const parsed = new Date(`${raw}T00:00:00Z`);
    if (Number.isNaN(parsed.getTime()) || parsed.toISOString().slice(0, 10) !== raw) {
      throw new Error('date answer must be a real ISO calendar date');
    }
    return raw;
  };

  if (value === null || value === undefined) return null;
  if (fieldType === 'text' || fieldType === 'textarea') {
    if (typeof value !== 'string') throw new Error('text-like answers must be strings');
    return nfc(value);
  }
  if (fieldType === 'email') {
    if (typeof value !== 'string') throw new Error('email answers must be strings');
    return nfc(value).trim();
  }
  if (fieldType === 'number') return canonicalDecimal(value);
  if (fieldType === 'date') return canonicalDate(value);
  if (fieldType === 'select' || fieldType === 'radio') {
    if (typeof value !== 'string') throw new Error('choice answers must be strings');
    const normalized = nfc(value).trim();
    if (!normalized) throw new Error('choice answer must be non-empty');
    return normalized;
  }
  if (fieldType === 'checkbox') {
    if (typeof value === 'boolean') return value;
    if (!Array.isArray(value)) throw new Error('checkbox answer must be boolean or array');
    const normalized = value.map((item) => {
      if (typeof item !== 'string') throw new Error('checkbox option answers must be strings');
      const option = nfc(item).trim();
      if (!option) throw new Error('checkbox option answers must be non-empty');
      return option;
    });
    return [...new Set(normalized)].sort();
  }
  if (fieldType === 'consent') {
    if (typeof value !== 'boolean') throw new Error('consent answer must be boolean');
    return value;
  }
  if (fieldType === 'file' || fieldType === 'unknown') {
    throw new Error(`${fieldType} answers are outside canonical model-visible payload identity`);
  }
  throw new Error(`unsupported form field type for canonical normalization: ${fieldType}`);
}

File: src/uexchanges/forms/normalization.py
from __future__ import annotations

import math
import unicodedata
from datetime import date, datetime
from decimal import Decimal, InvalidOperation
from typing import Any

from .models import FieldOwnership, FormField, FormFieldType


def _nfc(value: str) -> str:
    return unicodedata.normalize("NFC", value)


def _normalized_text(value: Any, *, trim: bool) -> str:
    if not isinstance(value, str):
        raise ValueError("text-like answers must be strings")
    normalized = _nfc(value.replace("\r\n", "\n").replace("\r", "\n"))
    return normalized.strip() if trim else normalized


def _normalized_number(value: Any) -> str:
    if isinstance(value, bool):
        raise ValueError("boolean is not a canonical number")
    if isinstance(value, float) and not math.isfinite(value):
        raise ValueError("number answers must be finite")
    if isinstance(value, Decimal):
        number = value
    elif isinstance(value, (int, float, str)):
        raw = str(value).strip()
        if not raw:
            raise ValueError("number answer must not be empty")
        try:
            number = Decimal(raw)
        except InvalidOperation as exc:
            raise ValueError("number answer is not a valid decimal") from exc
    else:
        raise ValueError("number answers must be int, float, Decimal or decimal string")
    if not number.is_finite():
        raise ValueError("number answers must be finite")
    if number == 0:
        return "0"
    return format(number.normalize(), "f")


def _normalized_date(value: Any) -> str:
    if isinstance(value, datetime):
        raise ValueError("datetime is not a canonical date answer")
    if isinstance(value, date):
        return value.isoformat()
    if not isinstance(value, str):
        raise ValueError("date answers must be date or ISO YYYY-MM-DD strings")
    raw = value.strip()
    try:
        parsed = date.fromisoformat(raw)
    except ValueError as exc:
        raise ValueError("date answer must use ISO YYYY-MM-DD") from exc
    return parsed.isoformat()


def _normalized_checkbox(value: Any) -> bool | list[str]:
    if isinstance(value, bool):
        return value
    if not isinstance(value, (list, tuple, set, frozenset)):
        raise ValueError("checkbox answer must be boolean or a collection of option strings")
    normalized: set[str] = set()
    for item in value:
        if not isinstance(item, str):
            raise ValueError("checkbox option answers must be strings")
        option = _normalized_text(item, trim=True)
        if not option:
            raise ValueError("checkbox option answers must be non-empty")
        normalized.add(option)
    return sorted(normalized)


def normalize_answer(field: FormField) -> Any:
    """Return the deterministic identity form of one model-visible answer.

    This function is for hashing/comparison identity, not for displaying or
    rewriting the user's answer. It never accepts a model-visible BLACK value.
    """
    if field.ownership is FieldOwnership.BLACK:
        if field.answer is not None:
            raise ValueError("BLACK answers must never enter canonical normalization")
        return None

    value = field.answer
    if value is None:
        return None

    if field.field_type in {FormFieldType.TEXT, FormFieldType.TEXTAREA}:
        return _normalized_text(value, trim=False)
    if field.field_type is FormFieldType.EMAIL:
        return _normalized_text(value, trim=True)
    if field.field_type is FormFieldType.NUMBER:
        return _normalized_number(value)
    if field.field_type is FormFieldType.DATE:
        return _normalized_date(value)
    if field.field_type in {FormFieldType.SELECT, FormFieldType.RADIO}:
        normalized = _normalized_text(value, trim=True)
        if not normalized:
            raise ValueError("choice answer must be non-empty")
        return normalized
    if field.field_type is FormFieldType.CHECKBOX:
        return _normalized_checkbox(value)
    if field.field_type is FormFieldType.CONSENT:
        if not isinstance(value, bool):
            raise ValueError("consent answer must be boolean")
        return value
    if field.field_type in {FormFieldType.FILE, FormFieldType.UNKNOWN}:
        raise ValueError(f"{field.field_type.value} answers are outside canonical model-visible payload identity")
    raise ValueError(f"unsupported form field type for canonical normalization: {field.field_type.value}")
